File: 2025/day02.py
VVfrom utils import get_data

def part1(filepath: str):
    data = get_data(filepath)[0].split(",")
    s = 0
    for istruzione in data:
        l,r = istruzione.split("-")
        l = int(l)
        r = int(r)
        # print(l,r)
        for n in range(l,r+1):
            lun = len(str(n))
            nStr = str(n)
            # print(lun)
            half = lun // 2
            # print(nStr[0:half+1],nStr[half:])
            if lun % 2 == 0:
                half = lun // 2
                if nStr[0:half] == nStr[half:]:
                    # print(n)
                    s += n
    return s

def part2(filepath: str):
    data = get_data(filepath)[0].split(",")
    res: list[int] = []
    for istruzione in data:
        l,r = istruzione.split("-")
        l = int(l)
        r = int(r)
        for n in range(l,r+1):
            lun = len(str(n))
            nStr = str(n)
            half = lun // 2
            for j in range(2, half+1):
                for k in range(2,half+1):
                    if nStr[0:j]*k == nStr:
                        res.append(n)
            # nSorted = "".join(sorted(nStr))
            if (nStr[0:half] == nStr[half:]) or (len(set(nStr)) == 1 and len(nStr) != 1):
                res.append(n)
    # print(set(res))
    return sum(set(res))


print(part1("./inputs/2.txt"))
print(part2("./inputs/2.txt"))

File: 2025/aoc_25/src/days/day02.rs
use crate::days::utils::get_data;

#[allow(unused)]
pub fn part1(filepath: &str) -> i64 {
    let binding = get_data(filepath);
    let data: Vec<&str> = binding[0].split(",").collect();
    let mut s = 0;
    for istruzione in data {
        let valori: Vec<&str> = istruzione.split("-").collect();
        // println!("{} ++ {}", valori[0], valori[1]);
        let l = valori[0].parse::<i64>().unwrap();
        let r = valori[1].parse::<i64>().unwrap();
        for n in l..r {
            let lun = n.to_string().len();
            let n_str = n.to_string();
            let half = lun / 2;
            if lun % 2 == 0 {
                if n_str[0..half] == n_str[half..] {
                    s += n
                }
            }
        }
    }
    s
}

pub fn part2(filepath: &str) -> i64 {
    let binding = get_data(filepath);
    let data: Vec<&str> = binding[0].split(",").collect();
    let mut ar: Vec<i64> = vec![];
    for istruzione in data {
        let valori: Vec<&str> = istruzione.split("-").collect();
        // println!("{} ++ {}", valori[0], valori[1]);
        let l = valori[0].parse::<i64>().unwrap();
        let r = valori[1].parse::<i64>().unwrap();
        for n in l..r {
            let lun = n.to_string().len();
            let n_str = n.to_string();
            let half = lun / 2;
            for j in 2..half + 1 {
                for k in 2..half + 1 {
                    if n_str[0..j].repeat(k) == n_str {
                        if !ar.contains(&n) {
                            ar.push(n);
                        }
                    }
                }
            }
            if lun % 2 == 0 {
                if n_str[0..half] == n_str[half..]
                    || n_str.chars().all(|c| c == n_str.chars().next().unwrap())
                {
                    if !ar.contains(&n) {
                        ar.push(n)
                    }
                }
            }
        }
    }
    ar.into_iter().sum()
}
